src/game.cpp

```diff
@@ -1214,20 +1214,22 @@ ReturnValue Game::internalMoveItem(Cylinder* fromCylinder, Cylinder* toCylinder,
 		}
 
 		int32_t newCount = m - n;
-		if (newCount > 0) {
+		if (newCount == item->getItemCount()) {
+			// full item is moved (move count is the same as item count)
+		} else if (newCount > 0) {
 			moveItem = item->clone();
 			moveItem->setItemCount(newCount);
 		} else {
+			if (item->isRemoved()) {
+				ReleaseItem(item);
+			}
+			
 			moveItem = nullptr;
 		}
```

Comment:
Assuming the new count can never be negative, how about this?

```suggestion
		assert(newCount >= 0);
		if (newCount != item->getItemCount()) {
			moveItem = item->clone();
			moveItem->setItemCount(newCount);
		} else if (newCount == 0) {
			if (item->isRemoved()) {
				ReleaseItem(item);
			}
			
			moveItem = nullptr;
		}
```

Reply:
Attention: the code above may cause duplication bugs that I couldn't trace. Please don't apply it without testing beforehand.